ral-facilities/datagateway-api
# Add DataGateway and PaNOSC Modes

This PR will close #256 

## Description
Adds a configuration option for `datagateway` and `panosc` modes by separating the config so that DataGateway API and Search API are objects. With the existing logic, it was complicated to make these objects optional so that if neither are present only the openapi endpoint is created for example. Because of this, I decided to use the [pydanic](https://pydantic-docs.helpmanual.io/) library which allows for data validation and settings management using type annotations. With the help of this library, these two objects are optional by default but if they are present then all of their fields are mandatory with exception of the optional fields in the `datagateway_api` object. This is because some fields here may need to be mandatory depending on what backend is used. For example, if `backend` is set to `python_icat` then the icat related config values must be present, therefore with the help of the `require_icat_config_value` validator function, these fields become mandatory. The app exits if the validation fails - i.e. a mandatory field is missing, a field is misspelled or a field value is of wrong data type.

With these changes in place, we can modify the relevant logic so that only the endpoints for DataGateway API and Search API are only created if their config objects are present in the `config.json` file, otherwise only the endpoint for the Open API is created. Realised that issue https://github.com/ral-facilities/datagateway-api/issues/283 was created with respect to this, so the above can be addressed as part of it.

## Testing Instructions
Unit tests have been written to test the config though there is no harm if the manual tests at the bottom are performed.
Note: The test that was creating a temporary file was failing (even without the config changes) due to Windows file permission issue. A unit test should run in isolation and should not touch the file system at all therefore I instead mocked the `builtins.open` to fake `with open` so there is no more need for a temp file to be created and read.

- [x] The application should exit if:
  - [x] the `db_url` in the `datagateway_api` object is missing in `config.json` when `backend` is set to `db`
  - [x] any or all of the icat related fields are missing in the `datagateway_api` object when `backend` is set to `python_icat`
  - [x] the `config.json` file cannot be found
- [x] The application should not exit if:
  - [x] the `datagateway_api` and `search_api` objects are missing in `config.json`.
  - [x] the `db_url` in the `datagateway_api` object is missing in `config.json` when `backend` is set to `python_icat`
  - [x] any or all of the icat related fields are missing in the `datagateway_api` object when `backend` is set to `db`

## Agile Board Tracking
Connect to #256
